### ISSUE TITLE
Save Modified Config File

### ISSUE THREAD

**Issue by p0fi** [guangie88/spdlog_setup#12]
Hi,

I came over from the original spdlog repo after I found this **awesome** extension for spdlog. After the removal of the dependencies its really much leaner and I am planning to use this in our production software. I extended spdlog myself before in a similar way, making it possible to load only log levels from a file. Being able to fully setup the loggers and sinks via config file is a very good improvement. The only thing my implementation provided which is missing now is the option to permanently change the config during runtime. Do you think a option like this could be implemented here also?

I could imagine a solution wich looks like this:

* The `default_config.toml` is never touched
* If the user modifies the log level during runtime and wants this change to persist he can save this.
* A new `custom_config.toml` is created and saved to disc. 
* When the program starts and tries to load the `default_config.toml` it looks for a `custom_config.toml` and prefers this over the `default_config.toml` if found.
* The user can also reset to `default_config.toml` at any time and the `custom_config.toml` is deleted. 
* A restart of the program would now load the `default_config.toml` since it cant find any other.

How do you think? Could you imagine such a feature?

Cheers!

**Comment by guangie88** [comment]
Hi, it is possible if there are API calls to perform all the getter methods (e.g. get all loggers, get logger name/format, etc.), but I am not currently sure if there are sufficient getter methods to achieve the above yet, since the current extension mostly uses the setter methods of `spdlog`.

I will check through the `spdlog` API, evaluate and let you know again.

**Comment by guangie88** [comment]
I am afraid I would have to decline your proposed suggestion; `spdlog` API has an extensive range of `setter` public methods, but very few `getter` methods (seems like only the name and level of a known logger can be retrieved). It would be impossible to properly implement your idea of persisting the registry of loggers (plus sinks and formatters) without the necessary `getter` methods, and at the same time I do not have control over the public `getter` methods required.

My apologies and I hope you can accept the explanation.

**Comment by p0fi** [comment]
Yeah I encountered pretty much the same issue when I was implementing it by myself.
Since I only dealt with the log levels and not with the sinks and formatters I implemented a getter for all loggers like this `std::vector<std::shared_ptr<spdlog::logger>> getAllLoggers();  `
I think something similar could be added to spdlog? Maybe we could figure out which getters are needed and add them to spdlog by making a PR?

**Comment by guangie88** [comment]
It could be difficult, basically all the fields that you see in the [static file config example](https://github.com/guangie88/spdlog_setup#static-file-configuration), which most of them pertain to specific `sinks`, and also the starting point `getAllLoggers` that you have mentioned (as well as `getSinks` and `getFormatters` per logger/sink), which pretty much means all `setters` must now have `getters` methods. From `spdlog`'s perspective, this would probably open up too much details for typical usage.

I am curious as to why you might need such a feature haha? Personally I haven't seen any log library doing the reverse (i.e. runtime logger fields back into a serialized config state), and haven't encountered issues with just having to set the config values before running the program.

**Comment by p0fi** [comment]
We have many distributed systems. Customers can request support and in such case its often necessary to increase log levels e.g.. Changing them during runtime is fine wich spdlog provides. But the customers are frequently restarting their devices, which would cause the default config to be loaded and the debugging log level is gone. So there needs to be a way to permanently save the log level. For this use case its all about the levels only. No need to change sinks or formatters.

**Comment by guangie88** [comment]
I see and thanks for sharing your use case. It is possible to implement the above with some form of basic override config file (similar to `docker-compose` override YML files?), and it does seem possible to have the override file following the original config file TOML syntax, just that in reality the override file is only able to have the `logger` and its overridden `level`.

For example based on my example original TOML config with `root` logger having `trace` level, one could override the level with the following `log_conf_linux.override.toml`:

```toml
[[logger]]
name = "root"
level = "info"
```

The above would make it possible to be forward compatible if in the future, when more `getter` methods become available (though unlikely).

There would also be a need to have the user specifying whether to allow override (override file will have a default filename based on the given original TOML file). As to when to perform the override persistence, I would suggest doing through an additional `spdlog_setup` method to take in the logger shared pointer, and the method will attempt to serialize whatever possible info based on the current `getter` methods available.

How does the above sound?

**Comment by p0fi** [comment]
That sounds like a very good solution for my given problem since I really only require the levels to be overridden. What do you think needs to be done and could I assist you in some way?

**Comment by guangie88** [comment]
No worries, I will try out a bit myself first since it shouldn't be too particularly difficult, then maybe you could provide some feedback for me 😃

**Comment by guangie88** [comment]
@p0fi Sorry for the delay, just delivered #14 and I believe it should be sufficient to resolve your issue.

There are 3 additional functions in the [v0.3.0-alpha.0](https://github.com/guangie88/spdlog_setup/tree/v0.3.0-alpha.0) that you can use:

```cpp
// override_toml_path can be your anticipated override file path to use
// the file can be missing at the start, which then only use base_toml_path
// returns true when override file is used (i.e. not missing / error opening file)
auto from_file_and_override(
    const std::string &base_toml_path, const std::string &override_toml_path)
    -> bool;

// only saves name + level, overwrite to always be false
// this allows the new logger entry to be saved into the new override file
// and subsequent saves should only change the level value
// from logger of the matching name
void save_logger_to_file(
    const std::shared_ptr<spdlog::logger> &logger,
    const std::string &toml_path,
    const bool overwrite = false);

// removes the TOML logger entry just for the matching name
// true if such entry exists and is deleted, otherwise false
auto delete_logger_in_file(
    const std::string &logger_name, const std::string &toml_path) -> bool;
```

Hopefully should be enough for your use case. Let me know if there are issues (hopefully not as new unit tests were added) or other comments. :slightly_smiling_face:

**Comment by p0fi** [comment]
Hey, 

thanks a lot! I tried this a little bit and it works nice when all is done in one source file (log setup and handling of overriding). I do want to handle the overrides based on user input, which is handled in a separate source file. When I now do in this source file `#include <spdlog_setup/conf.h>` I cant get it compiled anymore. I do get linker errors about duplicate symbols. I checked this a little and found that some non inline functions in the header files might be the reason. See [this](https://stackoverflow.com/questions/1787822/why-arent-include-guards-or-pragma-once-working) answer on stackoverflow. Can you confirm such problems?

**Comment by guangie88** [comment]
There are indeed missing `inline` for the `setup_error` methods, thanks for pointing out. I have patched them and can you help to check if my latest merged PR #15 resolves the issue?

You can just check out tag `v0.3.0-alpha.1` or the latest commit from `master`.

**Comment by p0fi** [comment]
Thanks for the fast patch! It works now!  
I'll think I will implement the new possibilities next week and could give you feedback after.
Thanks a lot again for implementing this feature in the first place! 👍
